src/Vector.cpp

```diff
@@ -0,0 +1,215 @@
+
+Vector::~Vector() {
+    if (_data != nullptr)
+        delete[] _data;
+}
+
+// Copy constructor
+Vector::Vector(const Vector& other)
+    : _size(0), _capacity(0), _data(nullptr) {
+    *this = other;
+}
+
+Vector& Vector::operator=(const Vector& other) {
+    if (this == &other)
+        return *this;
+
+    for (int i = 0; i < other.size(); i++) {
+        push_back(other[i]);
+    }
+    return *this;
+}
+
+// Move Constructor
+Vector::Vector(Vector&& other) noexcept
+    : _size(0), _capacity(0), _data(nullptr)
+{
+    if (this != &other) {
+        swap(other);
+    }
+}
+
+// Move Assignment Operator
+Vector& Vector::operator=(Vector&& other) noexcept
+{
+    if (this != &other) {
+        swap(other);
+    }
+    return *this;
+}
+
+// Access operator
+int& Vector::operator[](int index) {
+    if (index < 0 || index >= _size) {
+        throw std::out_of_range("Index out of bounds");
+    }
+    return _data[index];
+}
+
+const int& Vector::operator[](int index) const {
+    if (index < 0 || index >= _size) {
+        throw std::out_of_range("Index out of bounds");
+    }
+    return _data[index];
+}
+
+int* Vector::data() const {
+    return _data;
+}
+
+bool Vector::empty() const {
+    return _size == 0;
+}
+
+size_t Vector::size() const {
+    return _size;
+}
+
+void Vector::reserve(const size_t& new_cap) {
+    if (new_cap > _capacity) {
+        int* new_data = new int[new_cap];
+        for (size_t i = 0; i < _size; i++) {
+            new_data[i] = _data[i];
+        }
+
+        if (_data != nullptr)
+            delete[] _data;
+        _data = new_data;
+        _capacity = new_cap;
+    }
+}
+
+size_t Vector::capacity() const {
+    return _capacity;
+}
+
+void Vector::clear() {
+    _size = 0;
+}
+
+void Vector::resize(const size_t& count, const int& value) {
+    if (count < _size) {
+        _size = count;
+    }
+    else if (count > _size) {
+        for (std::size_t i = _size; i < count; i++) {
+            push_back(value);
+        }
+    }
+}
+
+void Vector::push_back(int value) {
+    if (_size >= _capacity) {
+        size_t new_capacity = (_capacity == 0) ? 1 : (_capacity < 128) ? _capacity * 2 : _capacity * 1.5;
+        reserve(new_capacity);
+    }
+    _data[_size++] = value;
+}
+
+void Vector::pop_back() {
+    if (_size > 0) {
+        _size--;
+    }
+}
+
+void Vector::swap(Vector& other) {
+    int* temp_data = _data;
+    size_t temp_size = _size;
+    size_t temp_capacity = _capacity;
+    _data = other._data;
+    _size = other._size;
+    _capacity = other._capacity;
+    other._data = temp_data;
+    other._size = temp_size;
+    other._capacity = temp_capacity;
+}
+
+void Vector::erase(size_t index) {
+    if (index < _size) {
+        // shift elements to the left
+        for (size_t i = index + 1; i < _size; i++) {
+            _data[i - 1] = _data[i];
+        }
+        _size--;
+    }
+}
+
+void Vector::insert(size_t index, int value) {
+    if (index <= _size) {
+        push_back(value);
```

Comment:
I understand you use `push_back()` here so it takes care of reserving extra space if needed, but this makes your code harder to understand because it's surprising to see you put the value at the end of the vector in an insert operation. 
Think about how you can make this code more readable